### PR TITLE
Build query at split level

### PR DESCRIPTION
**Pull request by evanxg852000**
### Context / purpose
Closes https://github.com/quickwit-inc/quickwit/issues/365

### Description
- Few changes to make sure the query is built at the split level
- Changed to make sure Field are accessed on the correct split schema

### How was this PR tested?
Create, Index, Search index

### Checklist
- [x] tested locally
- [ ] added unit tests
- [ ] included documentation

### CODE REVIEW

**Comment by fmassot** [comment]
Nice work here. One question though: why not building the collector at the split level too? So we pass index_config and request down to the split level and we are sure to be consistent. Moreover, I don't like pattern like "build first collector without timestamp" and adding it later. 
I think this will make it easier to refactor, what do you think?

File: quickwit-index-config/src/config.rs [hunk 1]
```diff
@@ -70,17 +70,24 @@ pub trait IndexConfig: Send + Sync + Debug + DynClone + 'static {
...
     /// Returns the schema.
+    ///
+    /// Considering schema evolution, splits within an index can have different schema
+    /// over time. The schema returned here represents the most up-to-date schema of the index.
     fn schema(&self) -> Schema;
     /// Returns the query.
-    fn query(&self, request: &SearchRequest) -> Result<Box<dyn Query>, QueryParserError>;
+    fn query(
+        &self,
+        schema: Schema,
```
Comment (fulmicoton) [inline quickwit-index-config/src/config.rs:80]: Rename split_schema and document that this is the schema that is specific to the targetted split, as opposed to current schema.

File: quickwit-index-config/src/config.rs [hunk 2]
```diff
@@ -70,17 +70,24 @@ pub trait IndexConfig: Send + Sync + Debug + DynClone + 'static {
...
+        request: &SearchRequest,
+    ) -> Result<Box<dyn Query>, QueryParserError>;
     /// Returns the default sort
     fn default_sort_by(&self) -> SortBy {
         SortBy::DocId
     }
     /// Returns the timestamp field.
-    fn timestamp_field(&self) -> Option<Field> {
+    fn timestamp_field(&self, schema: &Schema) -> Option<Field> {
```
Comment (fulmicoton) [inline quickwit-index-config/src/config.rs:88]: same as above.

Comment (fulmicoton) [reply quickwit-index-config/src/config.rs:88]: I think we could remove the method entirely, what do you think?

File: quickwit-index-config/src/all_flatten_config.rs [hunk 3]
```diff
@@ -78,9 +78,13 @@ impl IndexConfig for AllFlattenIndexConfig {
         Ok(document)
     }
 
-    fn query(&self, request: &SearchRequest) -> Result<Box<dyn Query>, QueryParserError> {
+    fn query(
+        &self,
+        schema: Schema,
+        request: &SearchRequest,
+    ) -> Result<Box<dyn Query>, QueryParserError> {
         let default_search_field_names = vec![SOURCE_FIELD_NAME.to_string()];
```
Comment (fulmicoton) [inline quickwit-index-config/src/all_flatten_config.rs:86]: Why vec!?

`&[SOURCE_FIELD_NAME.to_string()]` works I think.

File: quickwit-search/src/collector.rs [hunk 4]
```diff
@@ -375,17 +368,36 @@ fn extract_fast_field_names(index_config: &dyn IndexConfig) -> Vec<String> {
 }
 
 /// Builds the QuickwitCollector, in function of the information that was requested by the user.
-pub fn make_collector(
+pub fn make_split_collector(
+    split_id: String,
     index_config: &dyn IndexConfig,
     search_request: &SearchRequest,
+    split_schema: &Schema,
```
Comment (fulmicoton) [inline quickwit-search/src/collector.rs:375]: :heart:

File: quickwit-search/src/collector.rs [hunk 5]
```diff
@@ -375,17 +368,36 @@ fn extract_fast_field_names(index_config: &dyn IndexConfig) -> Vec<String> {
 }
 
 /// Builds the QuickwitCollector, in function of the information that was requested by the user.
-pub fn make_collector(
+pub fn make_split_collector(
```
Comment (fulmicoton) [inline quickwit-search/src/collector.rs:371]: great idea to rename it. I'd name it `make_collector_for_split`

**Review by fulmicoton — APPROVED** [review]
Great job. I only had a few minor comments.

File: quickwit-search/src/search_stream/leaf.rs [hunk 6]
```diff
@@ -38,8 +41,8 @@ use tokio_stream::wrappers::UnboundedReceiverStream;
 // to tonic::Status as tonic::Status is required by the stream result
 // signature defined by proto generated code.
 pub async fn leaf_search_stream(
-    query: Box<dyn Query>,
-    fast_field_collector_builder: FastFieldCollectorBuilder,
+    index_config: Box<dyn IndexConfig>,
+    request: &SearchStreamRequest,
     split_ids: Vec<String>,
     storage: Arc<dyn Storage>,
     output_format: OutputFormat,
```
Comment (fmassot) [inline quickwit-search/src/search_stream/leaf.rs:48]: I think we can remove `output_format` param as it is in the `SearchStreamRequest`

File: quickwit-search/src/search_stream/leaf.rs [hunk 7]
```diff
@@ -48,13 +51,13 @@ pub async fn leaf_search_stream(
...
         let result_sender_clone = result_sender.clone();
+        let request_clone = request.clone();
         tokio::spawn(async move {
             let leaf_split_result = leaf_search_stream_single_split(
-                query_clone,
-                collector_for_split,
+                index_config_clone,
+                &request_clone,
                 split_storage,
                 output_format,
```
Comment (fmassot) [inline quickwit-search/src/search_stream/leaf.rs:62]: same remark on `output_format`

**Review by fmassot — APPROVED** [review]
only two minor changes to do, your PR is clearly improving the code :)

**Pull request merged**